### DISCUSSION TITLE
battle game

### DISCUSSION THREAD

**Discussion by ksf0** [coccoinomane/crabada.py#120] [General]
Any thoughts on making a new bot for battle game? I'm a novice coder currently but learning quickly and I'd love to help!

I know its offchain so it would mostly be mostly vision detection to my knowledge

**Comment by coccoinomane** [comment]
Thank you @ksf0 for your offer to help, and welcome to the community 💪

I agree that Battle Game cannot be automated on-chain. Also, by moving off-chain, the team has implicitly made an important statement: **Crabada is not a permissionless game**.

It makes a lot of sense, if you think about it: you can't have a game with economic incentives which is both fun and permissionless, because the bots will ruin the fun. We can see this as the **Play-to-Earn Trilemma** 😉

When I started Crabada.py, my intent was to **learn blockchain scripting**; by making it open source, I wanted to give back to the Crabada's community.

Along the way, I met great "partners in crime" who made the project flourish well beyond my expectations: @yigitest, who lately has been contributing more than me 🙏,, @EugeneTeu who made it possible to quickly implement Swimmer, @nikohd12 for his precious feedback, @loayei, @alisalama, @ck0111, @coinmasterlisting, @Mrooie, @JorgePadilla, @mertf, and many more.

Both of my original motivations (learning blockchain + giving back to the community) are now gone: Crabada is not a blockchain game anymore and, being permissioned, botting it would likely incur in bans from the team and, ultimately, hurt the community.

#### Where does this leave us?

For the reasons explained above, I am happy to support anybody who wants to improve the capabilities of Crabada.py via a blockchain approach, if that is possible.

However, I am not going to actively participate in attempts to bot the game via vision detection, click simulations, etc.

PS: I am considering making a separate project with the **Web3 client** embedded in Crabada.py, because there might be interest for a user-friendly wrapper of web3.py, preconfigured to work out-of-the-box with the most popular blockchains. Let me know if you are interested in participating in any form 🙂

Cheers,
Cocco

**Reply by ksf0** [reply]
Makes perfect sense. I didn't have knowledge of the team saying that. I was planning on doing it alone as a way to learn vision detection like opencv but decided to ask here if there were plans to. I'll have to find another project to use vision detection with then. Thank you for all your hard work and making this open source in the first place!

**Reply by iTopher-acr** [reply]
I am testing an extension that works roughly as follows:

1) check for a private key in AWS Secret Manager
2) if no key is there, check for a key stored in a local instance of a vault based on vaultproject.io
3) if no key is there, check the .env file

there will be a configuration parameter added to the .env file indicating which path to take

happy to formally include it at some point provided that I can get it working

**Reply by coccoinomane** [reply]
Hi @imcapplecrypto !

Have you checked out [this PR](https://github.com/coccoinomane/crabada.py/wiki/%F0%9F%94%93-Store-private-keys-remotely)? It already has a lot of the required logic.

PS: I think you might have written in the wrong discussion ;-)

Cheers,
Cocco

**Comment by iTopher-acr** [comment]
Vladimir and I are working together on this. I didn’t know he’d already
made an official pull request.

Chris.
